src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -1135,7 +1135,7 @@ void ObjectMonitor::UnlinkAfterAcquire(JavaThread* current, ObjectWaiter* curren
 void ObjectMonitor::exit(JavaThread* current, bool not_suspended) {
   void* cur = owner_raw();
   if (current != cur) {
-    if (current->is_lock_owned((address)cur)) {
+    if (!UseFastLocking && current->is_lock_owned((address)cur)) {
       assert(_recursions == 0, "invariant");
       set_owner_from_BasicLock(cur, current);  // Convert from BasicLock* to Thread*.
       _recursions = 0;
```

Comment:
With fast-locking, OM::exit() must not be called as long as the owner is ANONYMOUS. The protocol is the following:
- T1 fast-locks object O
- T2 attempts to lock object O, observes that it is already locked and inflates with ANONYMOUS owner, because it doesn't know who owns the monitor. It then gets in line to acquire the monitor.
- When T1 exits, it observes the monitor being owned by ANONYMOUS, and sets the owner to itself, and calls OM::exit(), thereby performing the regular hand-over to T2.
- That check for ANONYMOUS owner also needs to be done in the C2 fast-path, and when that happens, it calls into runtime to do the necessary hand-over.